src/bytes_mut.rs

```diff
@@ -602,65 +533,60 @@ impl BytesMut {
+            let mut original_capacity = 0;
+
+            if !shared.is_null() {
+                original_capacity = (*shared).capacity;
 
-            // First, try to reclaim the buffer. This is possible if the current
-            // handle is the only outstanding handle pointing to the buffer.
-            if (*shared).is_unique() {
-                // This is the only handle to the buffer. It can be reclaimed.
+                // First, try to reclaim the buffer. This is possible if the current
+                // handle is the only outstanding handle pointing to the buffer.
                 // However, before doing the work of copying data, check to make
                 // sure that the vector has enough capacity.
-                let v = &mut (*shared).vec;
-
-                if v.capacity() >= new_cap {
+                if (*shared).is_unique() && (*shared).capacity >= new_cap {
                     // The capacity is sufficient, reclaim the buffer
-                    let ptr = v.as_mut_ptr();
+                    let ptr = (*shared).data_ptr_mut();
 
-                    ptr::copy(self.ptr.as_ptr(), ptr, len);
+                    ptr::copy(self.ptr.as_ptr(), ptr.as_ptr(), len);
 
-                    self.ptr = vptr(ptr);
-                    self.cap = v.capacity();
+                    self.ptr = ptr;
+                    self.cap = (*shared).capacity;
 
                     return;
                 }
-
-                // The vector capacity is not sufficient. The reserve request is
-                // asking for more than the initial buffer capacity. Allocate more
-                // than requested if `new_cap` is not much bigger than the current
-                // capacity.
-                //
-                // There are some situations, using `reserve_exact` that the
-                // buffer capacity could be below `original_capacity`, so do a
-                // check.
-                let double = v.capacity().checked_shl(1).unwrap_or(new_cap);
-
-                new_cap = cmp::max(cmp::max(double, new_cap), original_capacity);
-            } else {
-                new_cap = cmp::max(new_cap, original_capacity);
             }
-        }
-
-        // Create a new vector to store the data
-        let mut v = ManuallyDrop::new(Vec::with_capacity(new_cap));
 
-        // Copy the bytes
-        v.extend_from_slice(self.as_ref());
+            // The backing storage capacity is not sufficient. The reserve request is
+            // asking for more than the initial buffer capacity. Allocate more
+            // than requested if `new_cap` is not much bigger than the current
+            // capacity.
+            //
+            // There are some situations, using `reserve_exact` that the
+            // buffer capacity could be below `original_capacity`, so do a
+            // check.
+            let double = self.cap.checked_shl(1).unwrap_or(new_cap);
```

Comment:
This is one part that leads to the failing tests.
The old version doubled the original capacity of the storage. However if that was split, the capacity of the split part could be much lower.

I'm not really sure whether going back to the original capacity or even doubling it makes sense => If someone split off a 100 byte chunk of a 1MB `BytesMut` and they want to write a few more bytes into it, it shouldn't grow towards 1 or even 2 MB - right?

I'm somehwat leaning towards ignoring the original capacity at all, just making decisions based on capacity of the current part. But maybe I'm just missing why it was added that way? Maybe to make sure that a receive buffer somewhere which uses `BytesMut` and from deserialized data is split off always stays at the same size? If people would want that, they could still `.reserve()` that size themselves after splitting.

Reply:
I agree that it seems like surprising behavior. I would vote to ignore the original capacity.